crates/libs/core/src/sync/cancel.rs

```diff
@@ -0,0 +1,729 @@
+        // than cancelled error.
+
+        // Try to receive the value from the sender
+        let innner =
+            <tokio::sync::oneshot::Receiver<T> as Future>::poll(Pin::new(&mut self.rx), _cx);
+        match innner {
+            Poll::Ready(x) => {
+                // error only happens when sender is dropped without sending.
+                match x {
+                    Ok(data) => Poll::Ready(Ok(data)),
+                    Err(_) => {
+                        if let Some(t) = self.token.as_ref() {
+                            if t.is_cancelled() {
+                                // cancel the SF ctx
+                                if let Err(e) = self.cancel_inner_ctx() {
+                                    return Poll::Ready(Err(e));
+                                }
+                                return Poll::Ready(Err(FabricErrorCode::OperationCanceled.into()));
+                            }
+                        }
+                        panic!("sender dropped without sending")
+                    }
+                }
+            }
+            Poll::Pending => {
+                // If the action is canceled we can safely stop and return canceled error.
+                if let Some(t) = &self.token {
+                    // this is cancel safe so we can poll it once and discard
+                    let fu = t.cancelled();
+                    let inner = std::pin::pin!(fu).poll(_cx);
+                    match inner {
+                        Poll::Ready(_) => {
+                            // operation cancelled
+                            if let Err(e) = self.cancel_inner_ctx() {
+                                return Poll::Ready(Err(e));
+                            }
+                            Poll::Ready(Err(FabricErrorCode::OperationCanceled.into()))
+                        }
+                        Poll::Pending => Poll::Pending,
+                    }
+                } else {
+                    Poll::Pending
+                }
+            }
+        }
+    }
+}
+
+pub struct FabricSender2<T> {
+    tx: tokio::sync::oneshot::Sender<T>,
+    token: Option<CancellationToken>,
+}
+
+impl<T> FabricSender2<T> {
+    fn new(
+        tx: tokio::sync::oneshot::Sender<T>,
+        token: Option<CancellationToken>,
+    ) -> FabricSender2<T> {
+        FabricSender2 { tx, token }
+    }
+
+    pub fn send(self, data: T) {
+        let e = self.tx.send(data);
+        if e.is_err() {
+            // In SF use case receiver should not be dropped by user.
+            // If it acctually dropped by user, it is ok to ignore because user
+            // does not want to want the value any more. But too bad SF has done
+            // the work to get the value.
+
+            // receiver should never be dropped if operation is not cancelled.
+            if let Some(t) = self.token {
+                if !t.is_cancelled() {
+                    panic!("receiver dropped.");
```

Comment:
panic!() or debug_assert!()? Panic!() feels a bit strong here, will the library/program malfunction if execution continues after this point?

Reply:
Changed to debug_assert. Practically this should not happen.